### ISSUE TITLE
Malformed input or input contains unmappable characters (regression in 6.4.0rc1)

### ISSUE THREAD

**Issue by novas0x2a** [bazelbuild/bazel#19798]
### Description of the bug:

I think I found a bit of a regression between 6.3.2 and 6.4.0rc1.

Here's the traceback:
```
(19:43:42) FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'REPOSITORY_DIRECTORY:@rules_pkg' (requested by nodes 'PACKAGE_LOOKUP:@rules_pkg//')
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:633)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:365)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: net.starlark.java.eval.Starlark$UncheckedEvalException: InvalidPathException thrown during Starlark evaluation (LOADING)
	at <starlark>.download_and_extract(<builtin>:0)
	at <starlark>._http_archive_impl(/home/build/.cache/bazel/_bazel_build/473e173c53a2f7270c35ce0fd77d7706/external/bazel_tools/tools/build_defs/repo/http.bzl:132)
Caused by: java.nio.file.InvalidPathException: Malformed input or input contains unmappable characters: /home/build/.cache/bazel/_bazel_build/473e173c53a2f7270c35ce0fd77d7706/external/rules_pkg/tests/testdata/utf8/sübdir
	at java.base/sun.nio.fs.UnixPath.encode(Unknown Source)
	at java.base/sun.nio.fs.UnixPath.<init>(Unknown Source)
	at java.base/sun.nio.fs.UnixFileSystem.getPath(Unknown Source)
	at java.base/java.nio.file.Path.of(Unknown Source)
	at java.base/java.nio.file.Paths.get(Unknown Source)
	at com.google.devtools.build.lib.vfs.JavaIoFileSystem.getNioPath(JavaIoFileSystem.java:81)
	at com.google.devtools.build.lib.vfs.JavaIoFileSystem.createDirectoryAndParents(JavaIoFileSystem.java:250)
	at com.google.devtools.build.lib.vfs.Path.createDirectoryAndParents(Path.java:460)
	at com.google.devtools.build.lib.bazel.repository.CompressedTarFunction.decompress(CompressedTarFunction.java:102)
	at com.google.devtools.build.lib.bazel.repository.DecompressorValue.decompress(DecompressorValue.java:134)
	at com.google.devtools.build.lib.bazel.repository.starlark.StarlarkBaseExternalContext.downloadAndExtract(StarlarkBaseExternalContext.java:720)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at net.starlark.java.eval.MethodDescriptor.call(MethodDescriptor.java:162)
	at net.starlark.java.eval.BuiltinFunction.fastcall(BuiltinFunction.java:77)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:638)
	at net.starlark.java.eval.Eval.evalCall(Eval.java:682)
	at net.starlark.java.eval.Eval.eval(Eval.java:497)
	at net.starlark.java.eval.Eval.execAssignment(Eval.java:109)
	at net.starlark.java.eval.Eval.exec(Eval.java:268)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:66)
	at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:173)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:638)
	at net.starlark.java.eval.Starlark.call(Starlark.java:604)
	at com.google.devtools.build.lib.bazel.repository.starlark.StarlarkRepositoryFunction.fetch(StarlarkRepositoryFunction.java:215)
	at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.fetchRepository(RepositoryDelegatorFunction.java:411)
	at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.compute(RepositoryDelegatorFunction.java:342)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:562)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:365)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
```

What I know so far:
- I've only reproduced on linux.
- It only seems to happen on arm64 (cannot seem to repro on amd64) (and yeah, agreed, this doesn't seem to make sense).
- It only happens on machines where the installed locale list is limited. If you add a `apt install locales-all` and a `bazel shutdown` (so the server sees the locales) the problem goes away.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Here's [a repro](https://github.com/novas0x2a/bazel-locale-repro). Just run `./run.sh`. (It requires a recent docker, linux, and if you're not on arm64 also requires binfmt-misc)

### Which operating system are you running Bazel on?

linux arm64

### What is the output of `bazel info release`?

release 6.4.0rc1

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

I tried to bisect (with bazelisk) and I got:
```
$ bazel --bisect=6.3.2..release-6.4.0rc1 build //...
--- Getting the list of commits between 6.3.2 and release-6.4.0rc1
The good Bazel commit is not an ancestor of the bad Bazel commit, overriding the good Bazel commit to the merge base commit 0f231ac8acabcd8aa309da041c98ab90a1552418
Found 78 commits between (0f231ac8acabcd8aa309da041c98ab90a1552418, release-6.4.0rc1]
--- Verifying if the given good Bazel commit (0f231ac8acabcd8aa309da041c98ab90a1552418) is actually good
2023/10/11 20:02:55 Using unreleased version at commit 0f231ac8acabcd8aa309da041c98ab90a1552418
2023/10/11 20:02:55 could not run Bazel: could not download Bazel: arm64 linux is unsupported
```

which is unfortunate.

It seems like maybe it's a regression of https://github.com/bazelbuild/bazel/issues/1653

**Comment by Wyverald** [comment]
Thanks for the report. Have you tested with rc2, by any chance?

@fmeum does this look like something that changed in 6.4.0?

**Comment by novas0x2a** [comment]
> Have you tested with rc2, by any chance?

I did; same as rc1, unfortunately.

**Comment by fmeum** [comment]
The only commit I'm aware of that touches this is my https://github.com/bazelbuild/bazel/commit/b90ab8baf507178853942e961450561475719f8e, but it is only in 6.4.0rc2. I will investigate further.

**Comment by fmeum** [comment]
@meteorcloudy Based on the stack trace, the native `UnixFileSystem` doesn't seem to be available and Bazel falls back to the standard Java file system. Is that expected on Linux arm64?

**Comment by fmeum** [comment]
The `String`s backing a `PathFragment` are generally expected to be "raw bytes encoded with Latin-1". The fallback `JavaIoFileSystem` doesn't seem to handle this, so it's probably going to cause issues when `sun.jnu.encoding` is not Latin-1 (https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/vfs/JavaIoFileSystem.java;l=69 as opposed to https://github.com/bazelbuild/bazel/blob/fe58af421bd4bc9163e8a7fb3553a047ace20a41/src/main/java/com/google/devtools/build/lib/unix/UnixFileSystem.java#L486).

@novas0x2a Could you give us a list of the locales that are installed on your machine when the issue occurs?

**Comment by fmeum** [comment]
@novas0x2a Could you build Bazel from source with https://github.com/bazelbuild/bazel/pull/19801 and check whether that fixes the issue? Ideally, could you also build Bazel from source at current master and verify that the issue persists?

**Comment by fmeum** [comment]
@meteorcloudy Many tests fail with the `JavaIoFileSystem`, including ones that affect incremental correctness (https://github.com/bazelbuild/bazel/pull/19800). What do you think of showing a warning on server startup when `JavaIoFileSystem` is used? We could also try to fix the bugs, but that may be quite a bit of effort.

**Comment by meteorcloudy** [comment]
> Based on the stack trace, the native UnixFileSystem doesn't seem to be available and Bazel falls back to the standard Java file system. Is that expected on Linux arm64?

I actually don't know, probably not. I suspect if it could be https://github.com/bazelbuild/bazel/commit/8ac1cc102b5e4c804bdf9c44c0aaf1c632d6076a where we enabled cc toolchain resolution and somehow built the arm64 binary different. This happened to Windows arm64 before.

**Comment by meteorcloudy** [comment]
This is how we currently build the Linux arm64 bazel binary: https://github.com/bazelbuild/continuous-integration/blob/7ac1c0cf5f381b9de993a3540c4de17f4751818e/pipelines/bazel-release.yml#L110-L112

basically running the same build commands on a Linux arm64 machine.

**Comment by meteorcloudy** [comment]
> Could you give us a list of the locales that are installed on your machine when the issue occurs?

From a working Linux arm64 docker image:
```
[root@ed9c2037856c repo]# locale
LANG=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=
[root@ed9c2037856c repo]# locale -a
C
POSIX
en_US.iso88591
```

From the provided image:
```
root@e99455c6e7c7:/usr/src# locale
LANG=
LANGUAGE=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=
root@e99455c6e7c7:/usr/src# locale -a
C
C.utf8
POSIX
```

**Comment by meteorcloudy** [comment]
OK, after some debugging, @fmeum and I realized it's due to blake3, and is already fixed by https://github.com/bazelbuild/bazel/commit/31812fe7689fc317afed4e80c383d01cfcbc7087

**Comment by meteorcloudy** [comment]
@iancha1992 @keertk Let's cherry-pick https://github.com/bazelbuild/bazel/commit/31812fe7689fc317afed4e80c383d01cfcbc7087 to fix this one.

**Comment by iancha1992** [comment]
@novas0x2a A fix for this issue has been included in [Bazel 6.4.0 RC3](https://releases.bazel.build/6.4.0/rc3/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!

**Comment by novas0x2a** [comment]
rc3 works for me, thanks a lot!
